lib/rubygems/ext/cargo_builder.rb

```diff
@@ -4,48 +4,67 @@
 # over the `cargo rustc` command which takes care of building Rust code in a way
 # that Ruby can use.
 class Gem::Ext::CargoBuilder < Gem::Ext::Builder
-  attr_reader :spec
+  attr_accessor :spec, :runner, :profile
 
   def initialize(spec)
+    require_relative "../command"
+    require_relative "cargo_builder/link_flag_converter"
+
     @spec = spec
+    @runner = self.class.method(:run)
+    @profile = :release
   end
 
   def build(_extension, dest_path, results, args = [], lib_dir = nil, cargo_dir = Dir.pwd)
-    require "rubygems/command"
     require "fileutils"
     require "shellwords"
 
-    build_crate(dest_path, results, args, cargo_dir)
-    ext_path = rename_cdylib_for_ruby_compatibility(dest_path)
-    finalize_directory(ext_path, dest_path, lib_dir, cargo_dir)
+    build_crate(_extension, dest_path, results, args, cargo_dir)
+    validate_cargo_build!(dest_path)
+    rename_cdylib_for_ruby_compatibility(dest_path)
+    finalize_directory(dest_path, lib_dir, cargo_dir)
     results
   end
 
-  private
+  def build_crate(_extension, dest_path, results, args, cargo_dir)
+    env = build_env
+    cmd = cargo_command(cargo_dir, dest_path, args)
+    runner.call cmd, results, 'cargo', cargo_dir, env
 
-  def build_crate(dest_path, results, args, cargo_dir)
-    manifest = File.join(cargo_dir, "Cargo.toml")
-
-    given_ruby_static = ENV["RUBY_STATIC"]
+    results
+  end
 
-    ENV["RUBY_STATIC"] = "true" if ruby_static? && !given_ruby_static
+  def build_env
+    build_env = rb_config_env
+    build_env["RUBY_STATIC"] = "true" if ruby_static? && ENV.key?('RUBY_STATIC')
+    build_env
+  end
 
+  def cargo_command(cargo_dir, dest_path, args = [])
```

Comment:
I want to expose the actual cargo command to run, so it can be used in a Makefile later.